_posts/2022-05-17-noise.md

```diff
@@ -0,0 +1,333 @@
+
+
+For more details on supported handshakes and on how these are implemented in Waku, we refer to [35/WAKU2-NOISE](https://rfc.vac.dev/spec/35/) RFC.
+
+
+
+### The `K1K1` Handshake
+
+If Alice and Bob know each others' static key (e.g., these are public or were already exchanged in a previous handshake) , they MAY execute a `K1K1` handshake.  In Noise notation *(Alice is on the left)* this can be sketched as:
+
+```
+ K1K1:
+    ->  s
+    <-  s
+       ...
+    ->  e
+    <-  e, ee, es
+    ->  se
+```
+
+We note that here only ephemeral keys are exchanged. This handshake is useful in case Alice needs to instantiate a new separate encrypted communication channel with Bob, e.g. opening multiple parallel connections, file transfers, etc.
+
+**Security considerations on identity-hiding (active attacker)**: no static key is transmitted, but an active attacker impersonating Alice can check candidates for Bob's static key.
+
+### The `XK1` Handshake
+
+Here, Alice knows how to initiate a communication with Bob and she knows his public static key: such discovery can be achieved, for example, through a publicly accessible register of users' static keys, smart contracts, or through a previous public/private advertisement of Bob's static key.
+
+A Noise handshake pattern that suits this scenario is `XK1`:
+
+```
+ XK1:
+    <-  s
+       ...
+    ->  e
+    <-  e, ee, es
+    ->  s, se
+```
+
+Within this handshake, Alice and Bob reciprocally authenticate their static keys `s` using ephemeral keys `e`. We note that while Bob's static key is assumed to be known to Alice (and hence is not transmitted), Alice's static key is sent to Bob encrypted with a key derived from both parties ephemeral keys and Bob's static key.
+
+**Security considerations on identity-hiding (active attacker)**: Alice's static key is encrypted with forward secrecy to an authenticated party. An active attacker initiating the handshake can check candidates for Bob's static key against recorded/accepted exchanged handshake messages.
+
+
+### The `XX` and `XXpsk0` Handshakes
+
+If Alice is not aware of any static key belonging to Bob (and neither Bob knows anything about Alice), she can execute an `XX` handshake, where each party tran**X**mits to the other its own static key. 
+
+The handshake goes as follows:
+
+```
+ XX:
+    ->  e
+    <-  e, ee, s, es
+    ->  s, se
+```
+
+We note that the main difference with `XK1` is that in second step Bob sends to Alice his own static key encrypted with a key obtained from an ephemeral-ephemeral Diffie-Hellman exchange.
+
+
+This handshake can be slightly changed in case both Alice and Bob pre-shares some secret `psk` which can be used to strengthen their mutual authentication during the handshake execution. One of the resulting protocol, called `XXpsk0`, goes as follow:
+
+```
+ XXpsk0:
+    ->  psk, e
+    <-  e, ee, s, es
+    ->  s, se
+```
```

Comment:
Why does this not use the ... notation introduced earlier, like so:

```
￼ XXpsk0:
￼    <-  psk
          ...
     ->  e
￼    <-  e, ee, s, es
￼    ->  s, se
```

Reply:
The `psk` is a special token and it may appear at multiple times of the handshake telling parties to MixKeyAndHash(psk) (for example to bind each step to some key, etc.). It is not allowed in a pre-message pattern by specification, which usually deal about previous knowledge of parties public keys: by being a preshared key material, in fact, it doesn't make sense that it appear with a direction <- or -> since is not a per-user information.

Reply:
Yes, I was a bit fast: in `-> psk, e` Alice first mixes psk, and then sends `e` (not encrypted, but she might have sent `s` and in this case `s` would have been encrypted with a key derived from `psk` - which is not recommended though), and in this sense it makes sense there's a direction since she's sending a message, that is `e`. If psk was in the pre-message, no message is exchanged, hence a direction doesn't make sense. Hope this clarifies!